go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,215 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	utilnet "k8s.io/utils/net"
+)
+
+type EgressDNS struct {
+	// Protects pdMap/namespaces operations
+	lock sync.Mutex
+	// holds DNS entries globally
+	dns *util.DNS
+	// this map holds dnsNames to the dnsEntries
+	dnsEntries map[string]*dnsEntry
+	// allows for the creation of addresssets
+	addressSetFactory AddressSetFactory
+
+	// Report change when Add operation is done
+	added          chan string
+	stopChan       chan struct{}
+	controllerStop <-chan struct{}
+}
+
+type dnsEntry struct {
+	//this map holds all the namespaces that a dnsName appears in
+	namespaces map[string]struct{}
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	// the addressSet that contains the current IPs
+	dnsAddressSet AddressSet
+}
+
+func NewEgressDNS(addressSetFactory AddressSetFactory, controllerStop <-chan struct{}) (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:               dnsInfo,
+		dnsEntries:        make(map[string]*dnsEntry),
+		addressSetFactory: addressSetFactory,
+
+		added:          make(chan string),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	go utilwait.Until(egressDNS.Sync, 0, egressDNS.stopChan)
```

Comment:
Actually I don't think we need utilwait.Until() at all here, since Sync() does the right thing and returns when it's stopped. I think we can just go `e.Sync()`.

Reply:
done